### ISSUE TITLE
How to improve the results in extreme unbalanced dataset?

### ISSUE THREAD

**Issue by TimbusCalin** [huggingface/setfit#161]
I am working on an extreme imbalanced dataset for the binary-classification task (somewhere around 99.2% class 0 and 0.8% class 2 therefore).

The problem is that I am training with `1000 examples per each class but the f1-score lies around 0.18, accuracy = 88%(as reported by trainer)`, by using the default:

```
loss_class=CosineSimilarityLoss,
batch_size = 32 (doubled from 16)
num_iterations = 20
num_epochs = 1
```

What parameters and to what extent should I vary them in order to achieve better results (I suspect the iterations and the epochs, but I don't know to what extent)?

BTW I do not have more data for class 1 therefore adding data for it is not possible.

**Comment by lewtun** [comment]
Hey @TimbusCalin thanks for trying out the library! If you have 1000 examples per class, I think you'll get much better results with standard fine-tuning, e.g. using `distilbert` or similar: https://huggingface.co/course/chapter3/3?fw=pt

To handle the imbalance, you could try either downsampling the majority class or weighting your loss function. Here's a video where I show how you can do the latter: https://www.youtube.com/watch?v=u--UVvH-LIQ&t=376s

SetFit mostly shines in the low-data regime with 8-16 examples / class - for anything above ~few hundred examples I think vanilla fine-tuning is usually better :)

**Comment by TimbusCalin** [comment]
I will have a look on the trainings you provided. I do use almost everyday huggingface so your suggestions are very (set)FIT for me :D. Thanks a lot!

**Comment by freecui** [comment]
how to weight CosineSimilarityLoss？@TimbusCalin
